### DISCUSSION TITLE
Building rqlite for webasm

### DISCUSSION THREAD

**Discussion by lardinator** [rqlite/rqlite#1137] [Q&A]
Hi!

I want to see if it's possible to use rqlite as an embedded database in the browser and use WebRTC to distribute the data. I saw people running Sqlite with webasm and wanted to try. Note – I am new to Go and webasm.

When I build rqlite for webassembler the `golang.org/x/sys/unix` isn't supported. I looked at the code and I think it's an import in the `cmd` package. Do you know if there is any workaround or maybe some parts of the `cmd` package can be omitted?

Any tips and hits greatly appreciated.

**Comment by otoolep** [comment]
I do not believe it's possible because rqlite is built on Go **and** C. During the build process the SQLite source code is compiled, and then linked with the Go code. C code is not compatible with webasm AFAIK.

rqlite would need a pure Go version of SQLite. Such code does exist (see https://gitlab.com/cznic/sqlite), but then rqlite would not be running the official SQLite code. Not ruling it out yet, but haven't tried it at this time.

**Reply by lardinator** [reply]
Thank you for answering!

I believe that it's possible to build C to wasm with [emscripten](https://emscripten.org). I believe Sqlite has started to port it to wasm [in this repo](https://sqlite.org/wasm/tree), more info [here](https://sqlite.org/wasm/doc/tip/about.md).

I think it's possible to port rqlite to wasm by compiling the Go and C code separately with these steps: 
1. Compile rqlite to object files.
2. Compile sqlite3 to object files.
3. Link the objects files with Emscripten and compile to wasm in the same order as the command `go build -x ./...` 

Regarding my previous question, do you have any idea where `golang.org/x/sys/unix` is used?
I tried to read though the commands from `go build -x ./...`.
The output is huge and will take some time for me to work through to find the library.  

Finally, could you explain how the C-code for sqlite3 is imported?
I saw an import for `github.com/rqlite/go-sqlite3 v1.27.0` in `go.mod` - is that it?

**Reply by lardinator** [reply]
I have done some digging regarding the imports.
I think `Bowery/prompt` imports `golang.org/x/sys/unix`.
Is the import Bowery/prompt a "must have" or a "nice to have" for rqlite?

**Comment by otoolep** [comment]
It's a requirement to build the command-line too, but not for `rqlited`, the actual server.

**Reply by lardinator** [reply]
Do you think it's practical to only interact with rqlite using `rqlite-js`?

That could be a workaround for me so I don't have to build the CLI.
